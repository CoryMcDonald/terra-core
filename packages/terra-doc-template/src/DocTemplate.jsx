import React from 'react';
import PropTypes from 'prop-types';
import PropsTable from 'terra-props-table';
import Markdown from 'terra-markdown';
import IndexExampleTemplate from './IndexPageTemplate';

const propTypes = {
  /**
   * The given component's package name
   */
  packageName: PropTypes.string,
  /**
   * The given component's readme file imported to a string
   */
  readme: PropTypes.string,
  /**
   * Holds objects with the following properties:
   * ```
   * title: The title of the example
   * description: A description of the example to be displayed below the title
   * example: The example to be displayed
   * source: The source code of the example
   * ```
   */
  examples: PropTypes.arrayOf(
    PropTypes.shape(
      {
        title: PropTypes.string,
        description: PropTypes.string,
        example: PropTypes.element,
        source: PropTypes.string,
      },
    ),
  ),
  /**
   * Holds objects with the following properties:
   * ```
   * componentSrc: The source code of the component
   * componentName: The name of the component
   * ```
   */
  propsTables: PropTypes.arrayOf(
    PropTypes.shape(
      {
        componentSrc: PropTypes.string,
        componentName: PropTypes.string,
      },
    ),
  ),
};

const defaultProps = {
  packageName: '',
  readme: '',
  examples: [],
  propsTables: [],
};

const DocTemplate = (props) => {
  const { packageName, readme, examples, propsTables } = props;
  let id = 0;

  // Assign unique identifiers to use as keys later
  for (let i = 0; i < examples.length; i += 1) {
    examples[i].id = id;
    id += 1;
  }

  for (let i = 0; i < propsTables.length; i += 1) {
    propsTables[i].id = id;
    id += 1;
  }

  const version = packageName ? `[![NPM version](http://img.shields.io/npm/v/${packageName}.svg)](https://www.npmjs.org/package/${packageName})` : undefined;

  return (
    <div>
      {version && <Markdown src={version} />}
      {readme && <Markdown src={readme} />}

      {examples.length > 0 && <h1 style={{ paddingBottom: '0.3em', borderBottom: '1px solid #eaecef' }}>Examples</h1>}
      {examples.map(example =>
        <IndexExampleTemplate
          title={example.title}
          example={example.example}
          exampleSrc={example.source}
          description={example.description}
          key={example.id}
        />,
      )}

      {propsTables.map(propsTable =>
        <PropsTable
          src={propsTable.componentSource}
          componentName={propsTable.componentName}
          key={propsTable.id}
        />,
      )}
    </div>
  );
};

DocTemplate.propTypes = propTypes;
DocTemplate.defaultProps = defaultProps;

export default DocTemplate;
