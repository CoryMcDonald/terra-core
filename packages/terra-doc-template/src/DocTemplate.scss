:local {
  .examples-header {
    padding-bottom: '0.3em';
    border-bottom: '1px solid #eaecef';
  }
}
