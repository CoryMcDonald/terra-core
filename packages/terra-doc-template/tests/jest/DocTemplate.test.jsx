import React from 'react';

import DocTemplate from '../../src/DocTemplate';
import readme from '../../examples/test-examples/TestReadme.md';
// eslint-disable-next-line import/no-duplicates
import TestComponentExample from '../../examples/test-examples/TestComponentExample';
// Had to remove !rawloader! from the next two imports as it broke jest, but in normal use it should be there
// eslint-disable-next-line import/no-webpack-loader-syntax, import/first, import/no-unresolved, import/extensions, import/no-duplicates
import exampleSrc from '../../examples/test-examples/TestComponentExample.jsx';
// eslint-disable-next-line import/no-webpack-loader-syntax, import/first, import/no-unresolved, import/extensions
import testComponentSrc from '../../examples/test-examples/TestComponent.jsx';

jest.mock('../../examples/test-examples/TestReadme.md', () => 'test-file-stub');
jest.mock('../../examples/test-examples/TestComponentExample.jsx', () => 'test-file-stub');
jest.mock('../../examples/test-examples/TestComponent.jsx', () => 'test-file-stub');
const exampleElement = <TestComponentExample />;
// Snapshot Tests
it('should render a default component with nothing', () => {
  const wrapper = shallow(<DocTemplate />);
  expect(wrapper).toMatchSnapshot();
});

it('should show the version', () => {
  const wrapper = shallow(<DocTemplate packageName="terra-doc-template" />);
  expect(wrapper).toMatchSnapshot();
});


it('should show the readme', () => {
  const wrapper = shallow(<DocTemplate readme={readme} />);
  expect(wrapper).toMatchSnapshot();
});

it('should show the version and readme', () => {
  const wrapper = shallow(<DocTemplate packageName="terra-doc-template" readme={readme} />);
  expect(wrapper).toMatchSnapshot();
});

it('should show one example', () => {
  const wrapper = shallow(<DocTemplate examples={[{ title: 'Test Example 1', description: 'Describing the test', example: exampleElement, source: exampleSrc }]} />);
  expect(wrapper).toMatchSnapshot();
});

it('should show multiple examples', () => {
  const wrapper = shallow(<DocTemplate
    examples={[{ title: 'Test Example 1', description: 'Describing the test', example: exampleElement, source: exampleSrc },
      { title: 'Test Example 2', description: 'Describing the test mk. 2', example: exampleElement, source: exampleSrc }]}
  />);
  expect(wrapper).toMatchSnapshot();
});

it('should show one props table', () => {
  const wrapper = shallow(<DocTemplate propsTables={[{ componentSource: testComponentSrc, componentName: 'Test Component' }]} />);
  expect(wrapper).toMatchSnapshot();
});

it('should show multiple props tables', () => {
  const wrapper = shallow(<DocTemplate
    propsTables={[{ componentSource: testComponentSrc, componentName: 'Test Component' },
      { componentSource: testComponentSrc, componentName: 'Test Component (Again)' }]}
  />);
  expect(wrapper).toMatchSnapshot();
});

it('should show the version, readme, examples, and props tables', () => {
  const wrapper = shallow(<DocTemplate
    packageName="terra-doc-template"
    readme={readme}
    examples={[{ title: 'Test Example 1', description: 'Describing the test', example: exampleElement, source: exampleSrc },
      { title: 'Test Example 2', description: 'Describing the test mk. 2', example: exampleElement, source: exampleSrc }]}
    propsTables={[{ componentSource: testComponentSrc, componentName: 'Test Component' },
      { componentSource: testComponentSrc, componentName: 'Test Component (Again)' }]}
  />);
  expect(wrapper).toMatchSnapshot();
});
