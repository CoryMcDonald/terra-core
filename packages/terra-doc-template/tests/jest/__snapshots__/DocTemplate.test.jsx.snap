// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test should show the version 1`] = `
<div>
  Hello
</div>
`;
